xds/src/main/java/io/grpc/xds/client/XdsClientImpl.java

```diff
@@ -747,4 +947,166 @@ private void notifyWatcher(ResourceWatcher<T> watcher, T update) {
     }
   }
 
+  private class ResponseHandler implements XdsResponseHandler {
+    final ServerInfo serverInfo;
+
+    ResponseHandler(ServerInfo serverInfo) {
+      this.serverInfo = serverInfo;
+    }
+
+    @Override
+    public void handleResourceResponse(
+        XdsResourceType<?> xdsResourceType, ServerInfo serverInfo, String versionInfo,
+        List<Any> resources, String nonce, ProcessingTracker processingTracker) {
+      checkNotNull(xdsResourceType, "xdsResourceType");
+      syncContext.throwIfNotInThisSynchronizationContext();
+      Set<String> toParseResourceNames =
+          xdsResourceType.shouldRetrieveResourceKeysForArgs()
+          ? getResourceKeys(xdsResourceType)
+          : null;
+      XdsResourceType.Args args = new XdsResourceType.Args(serverInfo, versionInfo, nonce,
+          bootstrapInfo, securityConfig, toParseResourceNames);
+      handleResourceUpdate(args, resources, xdsResourceType, processingTracker);
+    }
+
+    @Override
+    public void handleStreamClosed(Status error) {
+      syncContext.throwIfNotInThisSynchronizationContext();
+      ControlPlaneClient cpcForThisStream = serverCpClientMap.get(serverInfo);
+      cleanUpResourceTimers(cpcForThisStream);
+
+      boolean hadError = false;
+      for (Map<String, ResourceSubscriber<? extends ResourceUpdate>> subscriberMap :
+          resourceSubscribers.values()) {
+        for (ResourceSubscriber<? extends ResourceUpdate> subscriber : subscriberMap.values()) {
+          if (!subscriber.hasResult() && cpcForThisStream.equals(subscriber.controlPlaneClient)) {
+            if (!hadError) {
+              logger.log(XdsLogLevel.WARNING, "ADS stream closed with error: {0}", error);
+              // try to fallback to lower priority control plane client
+              if (doFallbackForAuthority(
+                  serverInfo, subscriber.serverInfos, subscriber.authority)) {
+                return;
+              }
+            }
+            hadError = true;
+            subscriber.onError(error, null);
+          }
+        }
+      }
+    }
+
+    @Override
+    public void handleStreamRestarted(ServerInfo serverInfo) {
+      if (isShutDown()) {
+        return;
+      }
+
+      syncContext.throwIfNotInThisSynchronizationContext();
+
+      for (Map<String, ResourceSubscriber<? extends ResourceUpdate>> subscriberMap :
+          resourceSubscribers.values()) {
+        for (ResourceSubscriber<? extends ResourceUpdate> subscriber : subscriberMap.values()) {
+          if (subscriber.controlPlaneClient != null
+              && subscriber.controlPlaneClient.getServerInfo().equals(serverInfo)) {
+            subscriber.restartTimer();
+          }
+        }
+      }
+    }
+
+    @Override
+    public void handleStreamReady(ServerInfo serverInfo) {
```

Comment:
Is the `onReady()` callback also supported by stubby xdsTransportFactory? @ejona86

Reply:
They have to have something that recognizes that a stream has come ready that can feed into these calls.